src/rgw/driver/d4n/d4n_directory.h

```diff
@@ -1,53 +1,101 @@
-#ifndef CEPH_RGWD4NDIRECTORY_H
-#define CEPH_RGWD4NDIRECTORY_H
+#ifndef CEPH_D4NDIRECTORY_H
+#define CEPH_D4NDIRECTORY_H
 
 #include "rgw_common.h"
 #include <cpp_redis/cpp_redis>
 #include <string>
 #include <iostream>
+#include <boost/lexical_cast.hpp>
 
-struct cache_obj {
-  std::string bucket_name; /* s3 bucket name */
-  std::string obj_name; /* s3 obj name */
+namespace rgw { namespace d4n {
+
+struct Address {
+  std::string host;
+  int port;
+};
+
+struct CacheObj {
+  std::string objName; /* S3 object name */
+  std::string bucketName; /* S3 bucket name */
+  time_t creationTime; /* Creation time of the S3 Object */
+  bool dirty;
+  std::vector<std::string> hostsList; /* List of hostnames <ip:port> of object locations for multiple backends */
 };
 
-struct cache_block {
-  cache_obj c_obj;
-  uint64_t size_in_bytes; /* block size_in_bytes */
-  std::vector<std::string> hosts_list; /* Currently not supported: list of hostnames <ip:port> of block locations */
+struct CacheBlock {
+  CacheObj cacheObj;
+  uint64_t blockId; /* block ID */
+  uint64_t size; /* Block size in bytes */
+  int globalWeight = 0;
+  std::vector<std::string> hostsList; /* List of hostnames <ip:port> of block locations */
 };
 
-class RGWDirectory {
+class Directory {
   public:
-    RGWDirectory() {}
-    CephContext *cct;
+    Directory() {}
+    CephContext* cct;
+};
+
+class ObjectDirectory: public Directory { // weave into write workflow -Sam
+  public:
+    ObjectDirectory() {}
+    ObjectDirectory(std::string host, int port) {
+      addr.host = host;
+      addr.port = port;
+    }
+
+    void init(CephContext* _cct) {
+      cct = _cct;
+      addr.host = cct->_conf->rgw_d4n_host;
```

Comment:
Will the object directory always have the same address as the block directory or should there be a config option for each directory type?